### PR TITLE
Clone nl parser

### PR DESCRIPTION
**Pull request by TEdwardJ**

### CODE REVIEW

File: src/main/java/com/vinylteam/vinyl/util/PriceUtils.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
+package com.vinylteam.vinyl.util;
+
+import com.vinylteam.vinyl.entity.Currency;
+import lombok.extern.slf4j.Slf4j;
+import org.jsoup.nodes.Document;
+
+import java.util.Optional;
+
+@Slf4j
+public class PriceUtils {
```
Comment (Greengrer) [inline src/main/java/com/vinylteam/vinyl/util/PriceUtils.java:10]: This class methods don't fit for other parsers. You need to indicate that this one is for juno parser

File: src/main/java/com/vinylteam/vinyl/util/PriceUtils.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+                log.debug("Got optional with currency from price details {'optionalCurrency':{}, 'priceDetails':{}}", optionalCurrency, priceDetails);
+                return optionalCurrency;
+            } catch (Exception e) {
+                log.error("Error while getting optional with currency from price details from link {'priceDetails':{}}", priceDetails);
+            }
+        }
+        log.warn("Can't find currency description from price details from link, returning empty optional {'priceDetails':{}}", fullPriceDetails);
+        return Optional.empty();
+    }
+}
```
Comment (Greengrer) [inline src/main/java/com/vinylteam/vinyl/util/PriceUtils.java:43]: Empty row here.

File: src/main/java/com/vinylteam/vinyl/util/impl/JunoVinylParser.java [hunk 3]
```diff
@@ -154,46 +155,23 @@ String getArtistFromDocument(Document document) {
...
-            }
+        if (fullPriceDetails.contains(" ")) {
+            fullPriceDetails = fullPriceDetails.substring(fullPriceDetails.lastIndexOf(" ") + 1);
         }
-        log.warn("Can't find currency description from price details from link, returning empty optional {'priceDetails':{}, 'link':{}}", fullPriceDetails, document.location());
-        return Optional.empty();
+        return PriceUtils.getCurrencyFromString(fullPriceDetails);
     }
 
+
```
Comment (Greengrer) [inline src/main/java/com/vinylteam/vinyl/util/impl/JunoVinylParser.java]: Remove extra empty row.

**Review by Greengrer — CHANGES_REQUESTED** [review]
Thank you for optimizing my messy code

File: src/main/java/com/vinylteam/vinyl/util/impl/CloneNlParser.java [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+import java.util.regex.Pattern;
+import java.util.stream.IntStream;
+
+import static java.util.stream.Collectors.toSet;
+
+@Slf4j
+public class CloneNlParser implements VinylParser {
+
+    //TODO: Tests for PriceUtils
+    protected static final String BASE_LINK = "https://clone.nl";
```
Comment (AlexKievDeveloper) [inline src/main/java/com/vinylteam/vinyl/util/impl/CloneNlParser.java:25]: Why protected?

File: src/main/java/com/vinylteam/vinyl/util/impl/CloneNlParser.java [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+        var fullListOfPageLinks =
+                IntStream.rangeClosed(1, maxPageNumber)
+                        .mapToObj(pageNumber -> pageLinkPattern.replaceAll(PAGE_NUMBER_PATTERN.toString(), "&page=" + pageNumber))
+                        .collect(toSet());
+        log.debug("Resulting set of page links (with no gaps in sequence) is {'pageLinks':{}}", pageLinks);
+        return fullListOfPageLinks;
+    }
+
+    Set<String> getAllGenreLinks() {
+        //TODO: "/all/tags" can contains vinyls too
```
Comment (AlexKievDeveloper) [inline src/main/java/com/vinylteam/vinyl/util/impl/CloneNlParser.java:116]: It is still needed?

File: src/main/java/com/vinylteam/vinyl/util/impl/CloneNlParser.java [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+        rawOffer.setArtist(artist);
+        rawOffer.setPrice(price);
+        rawOffer.setCurrency(priceCurrency);
+        rawOffer.setOfferLink(offerLink);
+        rawOffer.setImageLink(imageLink);
+        rawOffer.setGenre(genre);
+        log.debug("New Raw Offer is Formed {'rawOffer': {}}", rawOffer);
+        return rawOffer;
+    }
+
```
Comment (AlexKievDeveloper) [inline src/main/java/com/vinylteam/vinyl/util/impl/CloneNlParser.java:160]: Remove one of 2 empty rows please

File: src/main/java/com/vinylteam/vinyl/util/impl/CloneNlParser.java (hunk 4 above)
Comment (TEdwardJ) [reply src/main/java/com/vinylteam/vinyl/util/impl/CloneNlParser.java:25]: It is required for test

File: src/main/java/com/vinylteam/vinyl/util/impl/CloneNlParser.java (hunk 5 above)
Comment (TEdwardJ) [reply src/main/java/com/vinylteam/vinyl/util/impl/CloneNlParser.java:116]: it is the question to check in the future

File: src/main/java/com/vinylteam/vinyl/Starter.java [hunk 7]
```diff
@@ -78,7 +79,7 @@ public static void main(String[] args) throws Exception {
 //UTIL, FILL IN DATABASE
         ShopsParser shopsParser = new ShopsParser();
         RawOffersSorter rawOffersSorter = new RawOffersSorter();
-        List<VinylParser> vinylParsers = List.of(new VinylUaParser(), new JunoVinylParser());
+        List<VinylParser> vinylParsers = List.of(new VinylUaParser(), new JunoVinylParser(), new CloneNlParser());
```
Comment (AlexKievDeveloper) [inline src/main/java/com/vinylteam/vinyl/Starter.java]: List<VinylParser> vinylParsers = List.of(new VinylUaParser(), new JunoVinylParser(), new DecksParser(), new CloneNlParser());

File: src/main/java/com/vinylteam/vinyl/util/impl/CloneNlParser.java [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+        var imageLink = getHighResImageLinkFromDocument(releaseElement);
+        var offerLink = getOfferLinkFromDocument(releaseElement);
+        var price = getPriceFromDocument(releaseElement);
+        var priceCurrency = getOptionalCurrencyFromDocument(releaseElement);
+        var artist = getArtistFromDocument(releaseElement);
+        var release = getReleaseFromDocument(releaseElement);
+        var genre = getGenreFromDocument(releaseElement);
+
+        var rawOffer = new RawOffer();
+        rawOffer.setShopId(4);
```
Comment (littlepony99) [inline src/main/java/com/vinylteam/vinyl/util/impl/CloneNlParser.java]: 4 to local constant plz

File: src/main/java/com/vinylteam/vinyl/util/PriceUtils.java (hunk 1 above)
Comment (TEdwardJ) [reply src/main/java/com/vinylteam/vinyl/util/PriceUtils.java:10]: Not only for juno parser, it is used in clone.nl parser too. It was the main idea to extract common methods for price string parsing

**Review by littlepony99 — APPROVED** [review]

File: src/main/java/com/vinylteam/vinyl/util/impl/CloneNlParser.java (hunk 4 above)
Comment (AlexKievDeveloper) [reply src/main/java/com/vinylteam/vinyl/util/impl/CloneNlParser.java:25]: Better use package private

**Review by AlexKievDeveloper — APPROVED** [review]

**Pull request merged**